gflownet/gflownet.py

```diff
@@ -591,43 +643,143 @@ def trajectorybalance_loss(self, it, batch):
+                data_dict = pickle.load(f)
+                states_term = data_dict["x"]
+        else:
+            raise NotImplementedError(
+                "data must be either a list of states or a path to a .pkl file."
+            )
+        assert len(states_term) < max_data
+        envs = []
+        for state_idx, x in enumerate(states_term):
+            for traj_idx in range(n_trajectories):
+                idx = int(max_data * state_idx + traj_idx)
+                env = self.env.copy().reset(idx)
+                env.set_state(x, done=True)
+                envs.append(env)
+        # Sample trajectories
+        actions = []
+        valids = []
+        max_iters = n_trajectories * max_iters_per_traj
+        iters = 0
+        while envs:
+            # Sample backward actions
+            actions = self.sample_actions(
+                envs,
+                batch,
+                backward=True,
+                no_random=True,
+                times=times,
+            )
+            # Update environments with sampled actions
+            envs, actions, valids = self.step(envs, actions, backward=True)
+            # Add to batch
+            batch.add_to_batch(envs, actions, valids, backward=True, train=True)
+            assert all(valids)
+            # Filter out finished trajectories
+            envs = [env for env in envs if not env.equal(env.state, env.source)]
+        # Prepare data structures to compute log probabilities
+        traj_ids = np.array([k for k in batch.trajectories])
+        data_indices = tlong(traj_ids // max_data, device=self.device)
+        traj_indices = tlong(traj_ids % max_data, device=self.device)
+        logprobs_f = torch.full(
+            (data_indices.max() + 1, traj_indices.max() + 1),
+            -torch.inf,
             dtype=self.float,
             device=self.device,
-        ).index_add_(0, traj_indices, logprobs_f)
-        # Backward trajectories
-        policy_output_b = self.backward_policy(states)
-        logprobs_b = self.env.get_logprobs(
-            policy_output_b, False, actions, parents, masks_b
         )
-        sumlogprobs_b = torch.zeros(
-            batch.get_n_trajectories(),
+        logprobs_b = torch.full(
+            (data_indices.max() + 1, traj_indices.max() + 1),
+            -torch.inf,
             dtype=self.float,
             device=self.device,
-        ).index_add_(0, traj_indices, logprobs_b)
-        # Trajectory balance loss
-        loss = (
-            (self.logZ.sum() + sumlogprobs_f - sumlogprobs_b - torch.log(rewards))
-            .pow(2)
-            .mean()
         )
-        return loss, loss, loss
+        traj_indices_mat = torch.full(
+            (data_indices.max() + 1, traj_indices.max() + 1),
+            -1,
+            dtype=torch.long,
+            device=self.device,
+        )
+        traj_indices_mat[data_indices, traj_indices] = traj_indices
+        n_trajs_per_sample, _ = traj_indices_mat.max(dim=1)
```

Comment:
Probably could be simplified as we should have the same number of trajectories for each sample provided as an argument to this function

Reply:
Yes, I will remove all this.

Reply:
Done [here](https://github.com/alexhernandezgarcia/gflownet/pull/167/commits/698ba2c48c439bb43190ca24f5c4e510d575ac19)